domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt

```diff
@@ -0,0 +1,112 @@
+import org.oppia.android.domain.platformparameter.PlatformParameterController
+import org.oppia.android.util.threading.BackgroundDispatcher
+import java.lang.IllegalArgumentException
+import javax.inject.Inject
+
+/** Worker class that fetches and caches the latest platform parameters from the remote service. */
+class PlatformParameterSyncUpWorker private constructor(
+  context: Context,
+  params: WorkerParameters,
+  private val platformParameterController: PlatformParameterController,
+  private val platformParameterService: PlatformParameterService,
+  private val oppiaLogger: OppiaLogger,
+  private val exceptionsController: ExceptionsController,
+  @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
+) : CoroutineWorker(context, params) {
+
+  companion object {
+    /** Exception message when the type of values received in the network response are not valid. */
+    const val INCORRECT_TYPE_EXCEPTION_MSG =
+      "Platform parameter value has incorrect data type, ie. other than String/Int/Boolean"
+
+    /** Exception message when there are no values received in the network response. */
+    const val EMPTY_RESPONSE_EXCEPTION_MSG = "Received an empty map in the network response"
+
+    /** A Tag for the logs that are associated with PlatformParameterSyncUpWorker. */
+    const val TAG = "PlatformParameterWorker.tag"
+
+    /** Value of worker-type associated with this PlatformParameterSyncUpWorker. */
+    const val PLATFORM_PARAMETER_WORKER = "platform_parameter_worker"
+
+    /** Key for passing the worker-type as a parameter to PlatformParameterSyncUpWorker. */
+    const val WORKER_TYPE_KEY = "worker_type_key"
+  }
+
+  override suspend fun doWork(): Result {
+    return when (inputData.getString(WORKER_TYPE_KEY)) {
+      PLATFORM_PARAMETER_WORKER -> withContext(backgroundDispatcher) { refreshPlatformParameters() }
+      else -> Result.failure()
+    }
+  }
+
+  // Parses a map of platform parameter values into a [List<PlatformParameter>]. If the parameters
+  // are not of type String, Int or Boolean this function fails with an [IllegalArgumentException].
+  private fun parseNetworkResponse(response: Map<String, Any>): List<PlatformParameter> {
+    val platformParameterList: MutableList<PlatformParameter> = mutableListOf()
+    for (entry in response.entries) {
+      val platformParameter = PlatformParameter.newBuilder().setName(entry.key)
+      when (val value = entry.value) {
+        is String -> platformParameter.string = value
+        is Int -> platformParameter.integer = value
+        is Boolean -> platformParameter.boolean = value
+        else -> throw IllegalArgumentException(INCORRECT_TYPE_EXCEPTION_MSG)
+      }
+      platformParameterList.add(platformParameter.build())
+    }
+    return platformParameterList
+  }
+
+  /** Extracts platform parameters from the remote service and stores them in the cache store */
+  private fun refreshPlatformParameters(): Result {
```

Comment:
Two problems with this thing.
1. refresh function is not doing any task in async format so suspend is not needed
2. the execute method over platformParameterService gives a warning message as can be seen in this image

![Untitled](https://user-images.githubusercontent.com/64064110/128905108-8a39aeda-7516-4bd6-b09f-efb3fc58f9ce.png)

Reply:
TIL but apparently we can change the Retrofit interface to generate a suspend function & return the direct value: https://stackoverflow.com/a/64139147. That's probably the best choice here (though you *may* run into testing issues--please let me know if that's the case).

Reply:
With your decription, this idea looks good but consider these points - 
- the `refreshPlatformParameter` is a normal function, this means that the control will not return to the `doWork` function untill and unless the execution of `refreshPlatformParameter` is completed.
- as we call the execute method over the `platformParameterService`, it will make sure that the network request is performed synchronously, this means that our `refreshPlatformParameter` function will be blocked to execute further untill and unless we get a response/failure.

So we actually don't need to use suspend for `refreshPlatformParameter` because the current code is working synchrounously only. If we will follow the suggestion to make the `getPlatformParameterByVersion` to be suspend and return a Response\<Map\<String,Any\>\>, then we will need to find a new way to test these suspend function calls in the `PlatformParameterServiceTest`, also we will need to have changes at the places were `MockPlatformParameterService` is getting used. 

I would say we can keep this thing as a completely new Issue to convert the network tests to use Response instead of Calls. (ie. Future Task) as the milestone deadline is coming closer and I have one more PR stacked over this PR, which needs to be merged.

Reply:
Thanks for looking into this. We do have #2951 tracking related items, so cc'ing it here since this is a useful context to capture.

That being said, I'm not sure that I fully understand your concerns. The fact that ``refreshPlatformParameter`` is a regular function seems a bit moot since it's called in a suspend function (doWork), so it seems like we could easily make it a suspend function. Regarding the blocking behavior, that's actually the problem I think we need to try and avoid. Coroutines are never supposed to be blocked.

That being said, this seems workable for the needs that we currently have & given we don't have better infrastructural support in the app to interop with Retrofit (yet).